contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol

```diff
@@ -163,27 +149,74 @@ contract LinkAvailableBalanceMonitor is ConfirmedOwner, Pausable, AutomationComp
     return targetsToFund;
   }
 
-  /// @notice Send funds to the targets provided.
-  /// @param targetAddresses the list of targets to fund
   function topUp(address[] memory targetAddresses) public whenNotPaused {
-    uint256 topUpAmount = s_topUpAmount;
-    uint256 stopIdx = targetAddresses.length;
-    uint256 numCanFund = LINK_TOKEN.balanceOf(address(this)) / topUpAmount;
-    stopIdx = numCanFund < stopIdx ? numCanFund : stopIdx;
-    for (uint256 idx = 0; idx < stopIdx; idx++) {
-      (bool exists, uint256 minBalance) = s_watchList.tryGet(targetAddresses[idx]);
-      if (!exists) {
-        emit TopUpBlocked(targetAddresses[idx]);
-        continue;
+    MonitoredAddress memory target;
+    uint256 localBalance = i_linkToken.balanceOf(address(this));
+    address[] memory targetsTopUp = new address[](targetAddresses.length);
+    for (uint256 idx = 0; idx < targetAddresses.length; idx++) {
+      address targetAddress = targetAddresses[idx];
+      target = s_targets[targetAddress];
+      (bool needsFunding, ) = _needsFunding(targetAddress, target.minBalance);
+      if (
+        target.topUpAmount > 0 && localBalance >= target.topUpAmount && !(targetAddress == address(0)) && needsFunding
+      ) {
+        targetsTopUp[idx] = targetAddress;
+        localBalance -= target.topUpAmount;
+      } else {
+        emit TopUpBlocked(targetAddress);
       }
-      (bool needsFunding, address target) = _needsFunding(targetAddresses[idx], minBalance);
-      if (!needsFunding) {
-        emit TopUpBlocked(targetAddresses[idx]);
+    }
+    for (uint256 idx = 0; idx < targetsTopUp.length; idx++) {
+      address targetAddress = targetsTopUp[idx];
+      if (targetAddress == address(0)) {
         continue;
       }
-      LINK_TOKEN.transfer(target, topUpAmount);
-      emit TopUpSucceeded(targetAddresses[idx]);
+      target = s_targets[targetAddress];
+      bool success = i_linkToken.transfer(targetAddress, target.topUpAmount);
+      if (success) {
+        target.lastTopUpTimestamp = uint56(block.timestamp);
+        emit TopUpSucceeded(targetAddress);
+      } else {
+        emit TopUpFailed(targetAddress);
+      }
+    }
+  }
+
+  /// @notice checks the target (could be direct target or IAggregatorProxy), and determines
+  /// if it is elligible for funding
+  /// @param targetAddress the target to check
+  /// @param minBalance minimum balance required for the target
+  /// @return bool whether the target needs funding or not
+  /// @return address the address of the contract needing funding
+  function _needsFunding(address targetAddress, uint256 minBalance) private view returns (bool, address) {
```

Comment:
afaict, none of the callers of `_needsFunding()` ever use the `minBalance` return value - could we get rid of that? If we can, then all of the `needsFunding()` calls could be inlined. So we could do this...

```sol
      if (_needsFunding(targetAddress, target.minBalance)) { ... }
```

... instead of this ...

```sol
      (bool needsFunding, ) = _needsFunding(targetAddress, target.minBalance);
      if (needsFunding) { ... }
```

Reply:
You're totally right, changing it.